File: Data Structures & Algorithms/search-for-word/submission-1.cpp
class Solution {
private:
    bool exists(vector<vector<char>>& board, vector<vector<bool>>& vis, string& word, int id, int i, int j){
        if(id==word.length()) return true;
        if(i<0||j<0||i>=board.size()||j>=board[0].size()||vis[i][j]||word[id]!=board[i][j]){
            return false;
        }
        vis[i][j]=true;
        bool ans=false;
        ans=ans|exists(board,vis,word,id+1,i-1,j);
        ans=ans|exists(board,vis,word,id+1,i,j-1);
        ans=ans|exists(board,vis,word,id+1,i+1,j);
        ans=ans|exists(board,vis,word,id+1,i,j+1);
        vis[i][j]=false;
        return ans;
    }
public:
    bool exist(vector<vector<char>>& board, string word) {
        int n=board.size();
        int m=board[0].size();
        vector<vector<bool>> vis(n,vector<bool>(m,false));
        bool ans=false;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                ans=ans|exists(board,vis,word,0,i,j);
            }
        }
        return ans;
    }
};

File: Data Structures & Algorithms/n-queens/submission-1.java
class Solution {
    private boolean isValidPlace(List<StringBuilder> board, int i, int j){
        int n=board.size();
        int m=board.get(0).length();
        int k=1;
        while(i-k>=0&&j-k>=0){
            if(board.get(i-k).charAt(j-k)=='Q'){
                return false;
            }
            k++;
        }
        k=1;
        while(i-k>=0&&j+k<m){
            if(board.get(i-k).charAt(j+k)=='Q'){
                return false;
            }
            k++;
        }
        k=1;
        while(i-k>=0){
            if(board.get(i-k).charAt(j)=='Q'){
                return false;
            }
            k++;
        }
        return true;
    }

    private boolean solve(List<StringBuilder> board, int row, int n, List<List<String>> fa){
        if(row==n){
            List<String> tempBoard=new ArrayList();
            for(StringBuilder x:board){
                tempBoard.add(x.toString());
            }
            fa.add(tempBoard);
            return true;
        }
        boolean ans=false;
        for(int i=0;i<board.get(row).length();i++){
            if(isValidPlace(board,row,i)){
                board.get(row).setCharAt(i,'Q');
                if(solve(board,row+1,n,fa)) ans=true;
                board.get(row).setCharAt(i,'.');
            }
        }
        return ans;
    }

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> fa=new ArrayList();
        List<StringBuilder> board=new ArrayList();
        StringBuilder ts=new StringBuilder();
        for(int i=0;i<n;i++){
            ts.append('.');
        }
        for(int i=0;i<n;i++){
            board.add(new StringBuilder(ts.toString()));
        }
        solve(board,0,n,fa);
        return fa;
    }
}
